src/devices/src/virtio/balloon/device.rs

```diff
@@ -256,58 +259,85 @@ impl Balloon {
         METRICS.balloon.inflate_count.inc();
 
         let queue = &mut self.queues[INFLATE_INDEX];
-        let mut pages = Vec::with_capacity(MAX_PAGES_IN_DESC);
+        // The pfn buffer index used during descriptor processing.
+        let mut pfn_buffer_idx = 0;
         let mut needs_interrupt = false;
 
-        while let Some(head) = queue.pop(&mem) {
-            let len = head.len;
-            if !head.is_write_only() && len % SIZE_OF_U32 as u32 == 0 {
-                for index in (0..len).step_by(SIZE_OF_U32) {
-                    let addr = head
-                        .addr
-                        .checked_add(index as u64)
-                        .ok_or(BalloonError::MalformedDescriptor)?;
-
-                    let page_frame_number = mem
-                        .read_obj::<u32>(addr)
-                        .map_err(|_| BalloonError::MalformedDescriptor)?;
-
-                    pages.push(page_frame_number);
+        // Loop until we consume the entire queue.
+        while !queue.is_empty(&mem) {
+            // Internal loop processes descriptors and acummulates the pfns in `pfn_buffer`.
+            // Breaks out when there is not enough space in `pfn_buffer` to completely process
+            // the next descriptor.
+            while let Some(head) = queue.pop(&mem) {
+                let len = head.len as usize;
+                let max_len = MAX_PAGES_IN_DESC * SIZE_OF_U32;
+
+                if !head.is_write_only() && len % SIZE_OF_U32 == 0 {
+                    // Check descriptor pfn count.
+                    if len > max_len {
+                        error!(
+                            "Inflate descriptor has bogus page count {} > {}, skipping.",
+                            len as usize / SIZE_OF_U32,
+                            MAX_PAGES_IN_DESC
+                        );
+
+                        // Skip descriptor.
+                        continue;
+                    }
+                    // Break loop if `pfn_buffer` will be overrun by adding all pfns from current desc.
+                    if MAX_PAGE_COMPACT_BUFFER - pfn_buffer_idx < len as usize / SIZE_OF_U32 {
+                        queue.undo_pop();
+                        break;
```

Comment:
This will break the inner loop but not the second too. It creates a soft-lockup because the outer loop never finishes/breaks.

Reply:
After the first loop is broken, in the outer loop we process the current pfn buffer and reset `pfn_buffer_idx` to 0 and resume desc processing.

Reply:
Nevermind, `pfn_buffer_idx` is reset to `0` at the end of the second loop. So this is fine.